sound/soc/codecs/rt715-sdca.c

```diff
@@ -88,77 +135,78 @@ static int rt715_sdca_set_amp_gain_put(struct snd_kcontrol *kcontrol,
-		if (val_l <= mc->shift) {
-			gain_l_val = ~gain_l_val;
-			gain_l_val += 1;
+	for (i = 0; i < 2; i++) {
+		gain_val =
+			rt715_sdca_vol_gain(ucontrol->value.integer.value[i], mc->max,
+				mc->shift);
+		ret = regmap_write(rt715->mbq_regmap, mc->reg + i, gain_val);
+		if (ret != 0) {
+			dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
+				mc->reg + i, gain_val);
+			return ret;
 		}
-		gain_l_val &= 0xffff;
 	}
 
-	/* R channel */
-	gain_r_val = ucontrol->value.integer.value[1];
-	if (gain_r_val > mc->max)
-		gain_r_val = mc->max;
-	val_r = gain_r_val;
-
-	if (mc->shift == 8) {
-		gain_r_val = (gain_r_val * 10) << mc->shift;
-	} else {
-		gain_r_val =
-			((abs(gain_r_val - mc->shift) * RT715_SDCA_DB_STEP) << 8) / 1000;
-		if (val_r <= mc->shift) {
-			gain_r_val = ~gain_r_val;
-			gain_r_val += 1;
-		}
-		gain_r_val &= 0xffff;
-	}
+	return 1;
+}
 
-	/* Lch*/
-	for (i = 0; i < loop_cnt; i++) {
-		ret = regmap_write(rt715->mbq_regmap, mc->reg + i * 2, gain_l_val);
+static int rt715_sdca_set_amp_gain_4ch_put(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	struct rt715_sdca_kcontrol_private *p =
+		(struct rt715_sdca_kcontrol_private *)kcontrol->private_value;
+	unsigned int reg_base = p->reg_base;
+	const unsigned int gain_sft = 0x2f;
+	unsigned int gain_val, i;
+	int ret;
+
+	for (i = 0; i < 4; i++) {
+		gain_val =
+			rt715_sdca_vol_gain(ucontrol->value.integer.value[i], p->max,
+				gain_sft);
+		ret = regmap_write(rt715->mbq_regmap, reg_base + i,
+				gain_val);
 		if (ret != 0) {
 			dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
-				mc->reg + i * 2, gain_l_val);
+				reg_base + i, gain_val);
 			return ret;
 		}
 	}
 
-	/* Rch */
-	for (i = 0; i < loop_cnt; i++) {
-		reg = (i == 3) ? (mc->rreg - 2) | BIT(15) : mc->rreg + i * 2;
-		ret = regmap_write(rt715->mbq_regmap, reg, gain_r_val);
+	return 1;
+}
+
+static int rt715_sdca_set_amp_gain_8ch_put(struct snd_kcontrol *kcontrol,
+		struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
+	struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
+	struct rt715_sdca_kcontrol_private *p =
+		(struct rt715_sdca_kcontrol_private *)kcontrol->private_value;
+	unsigned int reg_base = p->reg_base, i;
+	const unsigned int gain_sft = 8;
+	unsigned int gain_val, reg;
+	int ret;
+
+	for (i = 0; i < 8; i++) {
+		gain_val =
+			rt715_sdca_boost_gain(ucontrol->value.integer.value[i], p->max,
+				gain_sft);
+		reg = i < 7 ? reg_base + i : (reg_base - 1) | BIT(15);
+		ret = regmap_write(rt715->mbq_regmap, reg, gain_val);
 		if (ret != 0) {
 			dev_err(component->dev, "Failed to write 0x%x=0x%x\n",
-				reg, gain_r_val);
+				reg, gain_val);
 			return ret;
 		}
 	}
 
-	return 0;
+	return 1;
```

Comment:
The value 1 should be returned only if something was really changed (if get callback returns another value). This is just blind "notify about all writes".